discovery_child_development/analysis/baseline_model.py

```diff
@@ -0,0 +1,423 @@
+                                   Defaults to "sub_category".
+        head (int, optional): The number of top combinations to return. Defaults to 20.
+
+    Returns:
+        Tuple[pd.Series, Union[Tuple, List]]: A tuple containing two elements:
+                                              1. A pandas Series containing the counts of the top combinations of labels.
+                                              2. The most frequent combination of labels, which can be a tuple or a list.
+
+    Note:
+        The type of the most frequent labels (tuple or list) returned depends on the form of the entries in df[label_col].
+    """
+    sub_category_combinations = df[label_col].value_counts()
+
+    top_combinations = sub_category_combinations.head(head)
+
+    labels = top_combinations.index[0]
+
+    return top_combinations, labels
+
+
+def find_most_common_row(df: pd.DataFrame) -> Tuple[str, int]:
+    """
+    Finds the most common row pattern in a one-hot-encoded dataframe. Each row is converted into a string representation,
+    and the function identifies the pattern that occurs most frequently, along with the count of its occurrences.
+
+    Args:
+        df (pd.DataFrame): The one-hot-encoded dataframe to analyze.
+
+    Returns:
+        Tuple[str, int]: A tuple containing two elements:
+                         1. A string representing the most common row pattern (eg '00000010000001').
+                         2. An integer indicating the number of times this pattern occurs in the dataframe.
+    """
+
+    # make a copy, otherwise this will modify the original dataframe even though we are not returning it
+    df_copy = df.copy()
+
+    # Convert each row into a string representation
+    df_copy["combined"] = df_copy.apply(
+        lambda row: "".join(row.astype(str).values), axis=1, result_type="reduce"
+    )
+
+    # Find the most common set of binary labels
+    most_common_set = df_copy["combined"].value_counts().idxmax()
+    # Find out how many times this combination occurs
+    count = df_copy["combined"].value_counts().max()
+
+    return most_common_set, count
+
+
+def get_label_probabilities(
+    df: pd.DataFrame, label_col: str = "sub_category", n: int = 1000, targets=None
+) -> pd.Series:
+    """Get label probabilities from long-form data
+
+    Args:
+        df (pd.DataFrame): a long-form dataframe with an ID column and a label column (where each ID can have multiple labels)
+        label_col (str, optional): The column that contains the label. Defaults to "sub_category".
+        n (int, optional): Denominator used when calculating the prevalence of labels. Should be the number of unique IDs. Defaults to 1000.
+        targets (_type_, optional): This should be used if you're calculating probabilities on the training set, but the training set does not contain all possible labels. Defaults to None.
```

Comment:
Hey, now that I look at it, this function is kind of messy, so let me know if you want me to rethink the process a bit, rather than just adding more info in the docstring.

The way this function is used in context is:
```
label_probabilities = get_label_probabilities(
            train_df,
            "sub_category",
            len(train_df["openalex_id"].unique()),
            targets=Y_train.columns,
        )
```
In this case, `train_df` was a long-form dataframe (one row per label) and did not contain every possible label. `Y` was created with a multilabel binarizer and later split into `Y_train` and `Y_test`, hence `Y_train` contains a column for every possible label.

I don't know if the process is robust... Is it data leakage if we're artificially adding labels from the test set to the training set like this? (Not that leakage is an issue for this kind of model anyway, but once it's a "real" model it would be a problem)

Reply:
No, I think it's OK to define every possible label beforehand.